Fixing various regressions

slavapestov opened the pull request:
While testing a user project against the master branch I found a large number of regressions. This PR collects fixes for them. A few I was not able to address but we're tracking them internally.

slavapestov committed:
Serialization: Serialize FuncDecl::hasForcedStaticDispatch()

Otherwise, a protocol conformance where the witness was a dynamic
property in another module would trigger an assertion while building
the materializeForSet witness, or miscompile and fail at runtime
if asserts are off.

slavapestov committed:
IRGen: Fix crash when emitting type nested inside constrained extension

getFormalTypeInContext() was using the context archetypes of the
original parent type, and not the extension.

slavapestov committed:
IRGen: Fix conversion of qualified Clang TypeDecls

... or something. This manifested as a crash when emitting an @objc thunk
for an NS_STRING_ENUM type.

Fixes <rdar://problem/35140812>.

rjmccall commented on lib/IRGen/GenCall.cpp:
```diff
@@ -866,10 +866,11 @@ llvm::Type *SignatureExpansion::expandExternalSignatureTypes() {
     auto clangTy = IGM.getClangType(param);
-    paramTys.push_back(clangTy);
+    auto canClangTy = IGM.getClangASTContext().getCanonicalParamType(clangTy);
+    paramTys.push_back(canClangTy);
```
I think the better fix here might be to call getUnqualifiedType in ClangTypeConverter::convert's code for getting the type out of a clang::TypeDecl.

slavapestov replied:
Ok, I'll take a look tomorrow.

slavapestov replied:
Ok, that made the change a bit simpler. PTAL

slavapestov commented:
Next reduced test case that needs fixing:

```
import Foundation

extension NSString {
  var x: Float { return 0.0 }
}

var str: String? = nil

func f() {
  _ = str?.x
}
```

And after this, there are 3 more bugs. Sigh...

slavapestov committed:
Sema: Coerce base to rvalue when calling a DeclViaBridging overload kind

This code path is only used for a special purpose hack to make
NSString members available on String. Since it is rarely
exercised, it was broken.

slavapestov committed:
AST: Fix a typo in a diagnostic

slavapestov committed:
AST: Don't include Token.h from Module.h

slavapestov committed:
Parse: __shared and __owned should be contextual keywords

This was a source compatibility regression, someone actually had
an identifier named __shared.

slavapestov commented:
@CodaFi You might be interested in the parser change for __shared and __owned.

slavapestov commented:
@swift-ci Please smoke test

jrose-apple commented on lib/AST/Module.cpp:
```diff
@@ -34,6 +34,7 @@
 #include "swift/AST/ProtocolConformance.h"
 #include "swift/Basic/Compiler.h"
 #include "swift/Basic/SourceManager.h"
+#include "swift/Parse/Token.h"
```
How is this better? Just fewer includes?

slavapestov replied:
Yeah, editing Token.h does not trigger recompilation of ~400 source files anymore.

slavapestov commented:
@swift-ci Please smoke test

slavapestov commented:
@swift-ci Please smoke test

slavapestov commented:
@swift-ci Please smoke test

slavapestov commented:
@swift-ci Please smoke test

slavapestov commented:
@swift-ci Please smoke test

slavapestov commented:
@swift-ci Please smoke test

CodaFi commented:
If `__shared` and `__owned` are contextual keywords, there shouldn't be anything stopping `inout` from being given the same treatment.

slavapestov commented:
@CodaFi inout has legacy behavior where it can appear before a parameter name. I think that's just a hard error now, so if we're OK with losing that, we could make it a contextual keyword. But as Jordan pointed out offline, contextual keywords make error recovery more difficult, so unless we feel there's something to be gained by allowing inout to be used as an identifier, I don't see any reason to do that. I also feel that if we had shared and owned as keywords from the start, they would not need to be contextual, but at this point in time we have source compatibility concerns. That's my only reason for changing this.

slavapestov commented:
@swift-ci Please smoke test

slavapestov commented:
@swift-ci Please smoke test

slavapestov commented:
@swift-ci Please smoke test Linux

slavapestov commented:
@shahmishal Any ideas what's going on here?

```
 > git --version # timeout=10
 > git fetch --tags --progress git@github.com:apple/swift.git +refs/pull/*:refs/remotes/origin/pr/*
ERROR: Timeout after 10 minutes
ERROR: Error fetching remote repo 'origin'
hudson.plugins.git.GitException: Failed to fetch from git@github.com:apple/swift.git
```

slavapestov commented:
@swift-ci Please smoke test Linux

slavapestov commented:
@swift-ci Please smoke test Linux
